<!--
  Copyright (C) 2025 Rostislav Suleimanov

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->

<h2>Why is this an issue?</h2>
<p>
    In Magento 2, using hardcoded (literal) namespace strings in PHP code can reduce flexibility and maintainability.
    Instead, dynamic approaches such ::class should be used to improve code
    robustness, prevent typos, and facilitate refactoring.
</p>

<h3>Noncompliant code example</h3>
<pre data-diff-id="2" data-diff-type="noncompliant">
    public function notOk()
    {
        $this->create('Magento\CustomerSegment\Model\Segment\Condition\Customer\Address'); // Noncompliant
    }
</pre>

<h3>Compliant example</h3>
<pre data-diff-id="1" data-diff-type="compliant">
  public function ok()
    {
        $this->create(Magento\CustomerSegment\Model\Segment\Condition\Customer\Address::class); // Compliant
    }
</pre>