<!--
  Copyright (C) 2025 Rostislav Suleimanov

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->

<h2>Why is this an issue?</h2>
<p>
    This rule enforces that Magento 2 HTML templates must not directly instantiate objects.
    All required objects must be passed from Block objects instead.
</p>
<p>
    In Magento's architecture, templates should only use objects that have been properly injected into and exposed by their associated Block classes.
    Direct object instantiation in templates violates Magento's dependency injection pattern, creates tight coupling, complicates testing, and reduces maintainability.
</p>

<h3>Noncompliant code example</h3>
<pre data-diff-id="2" data-diff-type="noncompliant">
$customer = new \Magento\Customer\Model\Customer(); // Noncompliant
</pre>

<h3>Compliant solution</h3>
<pre data-diff-id="1" data-diff-type="compliant">
$customer = $this->getCustomer(); // Compliant
</pre>

<h2>References</h2>
<ul>
    <li>
        <a href="https://developer.adobe.com/commerce/frontend-core/guide/templates/" target="_blank">
            Magento 2 Templates documentation
        </a>
    </li>
</ul>