<!--
  Copyright (C) 2025 Rostislav Suleimanov

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU Affero General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
-->

<h2>Why is this an issue?</h2>
<p>
    The rule "Explicit return types MUST BE declared on functions" is a coding standard guideline often used in modern PHP development,
    especially with strict typing enabled.
</p>
<p>It means that every function or method you define should specify the type of value that it will return.</p>
<p>This helps with code clarity, maintenance, and avoids potential runtime errors due to mismatched types.</p>

<h3>Noncompliant code example</h3>
<pre data-diff-id="2" data-diff-type="noncompliant">
function doSomething(){ // Noncompliant
    return 'test';
}
</pre>

<h3>Compliant solution</h3>
<pre data-diff-id="1" data-diff-type="compliant">
function doSomething() :string // compilant
{
    return 'test';
}
</pre>


<h3>Exceptions</h3>
<p>Magic methods (<code>__construct()</code>, <code>__destruct()</code>,
    <code>__call()</code>, <code>__callStatic()</code>, <code>__get()</code>, <code>__set()</code>, <code>__isset()</code>, <code>__unset()</code>,
    <code>__sleep()</code>, <code>__wakeup()</code>, <code>__toString()</code>, <code>__invoke()</code>, <code>__set_state()</code>,
    <code>__clone()</code>, <code>__debugInfo()</code>) are ignored.</p>
<pre>
function __construct(){...} // Compliant by exception
function __destruct(){...}  // Compliant by exception
</pre>

<h2>References</h2>
<ul>
    <li>
        <a href="https://www.php.net/manual/en/language.types.declarations.php" target="_blank">
            PHP Type declarations documentation
        </a>
    </li>
</ul>